diff --git a/document/js-api/index.bs b/document/js-api/index.bs
index c00cadeed..03e69e7e2 100644
--- a/document/js-api/index.bs
+++ b/document/js-api/index.bs
@@ -311,7 +311,8 @@ Note:
 
 <div algorithm>
   To <dfn>compile a WebAssembly module</dfn> from source bytes |bytes|, perform the following steps:
-    1. Let |module| be [=module_decode=](|bytes|). If |module| is [=error=], return [=error=].
+    1. Let |module| be [=module_decode=](|bytes|).
+    1. If |module| is [=error=], return [=error=].
     1. If [=module_validate=](|module|) is [=error=], return [=error=].
     1. Return |module|.
 </div>
@@ -332,7 +333,7 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
 <div algorithm>
   To <dfn>construct a WebAssembly module object</dfn> from a module |module| and source bytes |bytes|, perform the following steps:
 
-    1. Let |moduleObject| be a new {{Module}} object.
+    1. Let |moduleObject| be a [=/new=] {{Module}} object.
     1. Set |moduleObject|.\[[Module]] to |module|.
     1. Set |moduleObject|.\[[Bytes]] to |bytes|.
     1. Return |moduleObject|.
@@ -344,12 +345,12 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
     1. Let |promise| be [=a new promise=].
     1. Run the following steps [=in parallel=]:
         1. [=compile a WebAssembly module|Compile the WebAssembly module=] |bytes| and store the result as |module|.
-        1. [=Queue a task=] to perform the following steps. If |taskSource| was provided, queue the task on that task source.
-            1. If |module| is [=error=], reject |promise| with a {{CompileError}} exception.
+        1. [=Queue a task=] to perform the following steps:
+            1. If |module| is [=error=], [=reject=] |promise| with a {{CompileError}} exception.
             1. Otherwise,
                 1. [=Construct a WebAssembly module object=] from |module| and |bytes|, and let |moduleObject| be the result.
                 1. [=Resolve=] |promise| with |moduleObject|.
-    1. Return |promise|.
+    1. Return |promise|.[[Promise]].
 </div>
 
 <div algorithm>
@@ -360,15 +361,16 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
 
 <div algorithm="read-the-imports">
   To <dfn>read the imports</dfn> from a WebAssembly module |module| from imports object |importObject|, perform the following steps:
-    1. If |module|.[=imports=] [=list/is empty|is not empty=], and |importObject| is undefined, throw a {{TypeError}} exception.
+    1. Let |tuples| be [=module_imports=](|module|).
+    1. If |tuples| [=list/is empty|is not empty=], and |importObject| is undefined, throw a {{TypeError}} exception.
     1. Let |imports| be « ».
-    1. [=list/iterate|For each=] (|moduleName|, |componentName|, |externtype|) of [=module_imports=](|module|),
+    1. [=list/iterate|For each=] (|moduleName|, |componentName|, |externtype|) of |tuples|,
         1. Let |o| be [=?=] [$Get$](|importObject|, |moduleName|).
         1. If |o| [=is not an Object=], throw a {{TypeError}} exception.
         1. Let |v| be [=?=] [$Get$](|o|, |componentName|).
         1. If |externtype| is of the form [=func=] |functype|,
             1. If [$IsCallable$](|v|) is false, throw a {{LinkError}} exception.
-            1. If |v| has a \[[FunctionAddress]] internal slot, and therefore is an [=Exported Function=],
+            1. If |v| has a \[[FunctionAddress]] internal slot,
                 1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
             1. Otherwise,
                 1. [=Create a host function=] from |v| and |functype|, and let |funcaddr| be the result.
@@ -383,7 +385,7 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
                     1. Throw a {{LinkError}} exception.
                 1. If |valtype| is [=v128=],
                     1. Throw a {{LinkError}} exception.
-                1. Let |value| be [=ToWebAssemblyValue=](|v|, |valtype|).
+                1. Let |value| be [=?=] [=ToWebAssemblyValue=](|v|, |valtype|).
                 1. Let |store| be the [=surrounding agent=]'s [=associated store=].
                 1. Let (|store|, |globaladdr|) be [=global_alloc=](|store|, [=const=] |valtype|, |value|).
                 1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
@@ -403,15 +405,12 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
             1. Let |externtable| be the [=external value=] [=external value|table=] |tableaddr|.
             1. [=list/Append=] |externtable| to |imports|.
     1. Return |imports|.
-
-Note: This algorithm only verifies the right kind of JavaScript values are passed.
-The verification of WebAssembly type requirements is deferred to the
-"[=instantiate the core of a WebAssembly module=]" algorithm.
+    Note: This algorithm only verifies the right kind of JavaScript values are passed. The verification of WebAssembly type requirements is deferred to the "[=instantiate the core of a WebAssembly module=]" algorithm.
 </div>
 
 <div algorithm>
   To <dfn>create an exports object</dfn> from a WebAssembly module |module| and instance |instance|, perform the following steps:
-    1. Let |exportsObject| be [=!=] [$OrdinaryObjectCreate$](null).
+    1. Let |exportsObject| be [$OrdinaryObjectCreate$](null).
     1. [=list/iterate|For each=] (|name|, |externtype|) of [=module_exports=](|module|),
         1. Let |externval| be [=instance_export=](|instance|, |name|).
         1. Assert: |externval| is not [=error=].
@@ -439,7 +438,7 @@ The verification of WebAssembly type requirements is deferred to the
         1. Assert: |status| is true.
 
         Note: the validity and uniqueness checks performed during [=WebAssembly module validation=] ensure that each property name is valid and no properties are defined twice.
-    1. Perform [=!=] [$SetIntegrityLevel$](|exportsObject|, `"frozen"`).
+    1. Perform [=!=] [$SetIntegrityLevel$](|exportsObject|, ~frozen~).
     1. Return |exportsObject|.
 </div>
 
@@ -449,6 +448,7 @@ The verification of WebAssembly type requirements is deferred to the
     1. [=Create an exports object=] from |module| and |instance| and let |exportsObject| be the result.
     1. Set |instanceObject|.\[[Instance]] to |instance|.
     1. Set |instanceObject|.\[[Exports]] to |exportsObject|.
+    1. Return ~unused~.
 </div>
 
 <div algorithm>
@@ -469,17 +469,23 @@ The verification of WebAssembly type requirements is deferred to the
     1. Let |promise| be [=a new promise=].
     1. Let |module| be |moduleObject|.\[[Module]].
     1. [=Read the imports=] of |module| with imports |importObject|, and let |imports| be the result.
-        If this operation throws an exception, catch it, [=reject=] |promise| with the exception, and return |promise|.
+    1. If |imports| is an abrupt completion,
+        1. [=Reject=] |promise| with |imports|.
+        1. Return |promise|.[[Promise]].
     1. Run the following steps [=in parallel=]:
         1. [=Queue a task=] to perform the following steps:
             Note: Implementation-specific work may be performed here.
             1.  [=Instantiate the core of a WebAssembly module=] |module| with |imports|, and let |instance| be the result.
-                If this throws an exception, catch it, [=reject=] |promise| with the exception, and terminate these substeps.
+            1.  If |instance| is an abrupt completion,
+                1. [=Reject=] |promise| with |instance|.
+                1. Return ~unused~.
             1.  Let |instanceObject| be a [=/new=] {{Instance}}.
             1.  [=initialize an instance object|Initialize=] |instanceObject| from |module| and |instance|.
-                If this throws an exception, catch it, [=reject=] |promise| with the exception, and terminate these substeps.
+            1.  If |instance| is an abrupt completion,
+                1. [=Reject=] |promise| with |instance|.
+                1. Return ~unused~.
             1. [=Resolve=] |promise| with |instanceObject|.
-    1. Return |promise|.
+    1. Return |promise|.[[Promise]].
 </div>
 
 <div algorithm>
@@ -488,14 +494,14 @@ The verification of WebAssembly type requirements is deferred to the
     1. Let |promise| be [=a new promise=].
     1. [=Upon fulfillment=] of |promiseOfModule| with value |module|:
         1. [=asynchronously instantiate a WebAssembly module|Instantiate the WebAssembly module=] |module| importing |importObject|, and let |innerPromise| be the result.
-        1. [=Upon fulfillment=] of |innerPromise| with value |instance|.
+        1. [=Upon fulfillment=] of |innerPromise| with value |instance|:
             1. Let |result| be the {{WebAssemblyInstantiatedSource}} value «[ "{{WebAssemblyInstantiatedSource/module}}" → |module|, "{{WebAssemblyInstantiatedSource/instance}}" → |instance| ]».
             1. [=Resolve=] |promise| with |result|.
         1. [=Upon rejection=] of |innerPromise| with reason |reason|:
             1. [=Reject=] |promise| with |reason|.
     1. [=Upon rejection=] of |promiseOfModule| with reason |reason|:
         1. [=Reject=] |promise| with |reason|.
-    1. Return |promise|.
+    1. Return |promise|.[[Promise]].
 </div>
 
 <div algorithm>
@@ -545,10 +551,10 @@ interface Module {
 
 <div algorithm>
   The <dfn>string value of the extern type</dfn> |type| is
-    * "function" if |type| is of the form [=func=] <var ignore>functype</var>
-    * "table" if |type| is of the form [=table=] <var ignore>tabletype</var>
-    * "memory" if |type| is of the form [=mem=] <var ignore>memtype</var>
-    * "global" if |type| is of the form [=global=] <var ignore>globaltype</var>
+    1. Return "function" if |type| is of the form [=func=] <var ignore>functype</var>
+    1. Return "table" if |type| is of the form [=table=] <var ignore>tabletype</var>
+    1. Return "memory" if |type| is of the form [=mem=] <var ignore>memtype</var>
+    1. Return "global" if |type| is of the form [=global=] <var ignore>globaltype</var>
 </div>
 
 <div algorithm>
@@ -593,8 +599,9 @@ interface Module {
     1. If |module| is [=error=], throw a {{CompileError}} exception.
     1. Set **this**.\[[Module]] to |module|.
     1. Set **this**.\[[Bytes]] to |stableBytes|.
+    1. Return ~unused~.
 
-Note: Some implementations enforce a size limitation on |bytes|. Use of this API is discouraged, in favor of asynchronous APIs.
+    Note: Some implementations enforce a size limitation on |bytes|. Use of this API is discouraged, in favor of asynchronous APIs.
 </div>
 
 <h3 id="instances">Instances</h3>
@@ -613,8 +620,9 @@ interface Instance {
     1. [=Read the imports=] of |module| with imports |importObject|, and let |imports| be the result.
     1. [=Instantiate the core of a WebAssembly module=] |module| with |imports|, and let |instance| be the result.
     1. [=initialize an instance object|Initialize=] **this** from |module| and |instance|.
+    1. Return ~unused~.
 
-Note: The use of this synchronous API is discouraged, as some implementations sometimes do long-running compilation work when instantiating.
+    Note: The use of this synchronous API is discouraged, as some implementations sometimes do long-running compilation work when instantiating.
 </div>
 
 <div algorithm>
@@ -667,7 +675,7 @@ which can be simultaneously referenced by multiple {{Instance}} objects. Each
     1. Let |buffer| be a new {{ArrayBuffer}} with the internal slots \[[ArrayBufferData]], \[[ArrayBufferByteLength]], \[[ArrayBufferMaxByteLength]], and \[[ArrayBufferDetachKey]].
     1. Set |buffer|.\[[ArrayBufferData]] to |block|.
     1. Set |buffer|.\[[ArrayBufferByteLength]] to |length|.
-    1. Set |buffer|.\[[ArrayBufferMaxByteLength]] is |maxsize|.
+    1. Set |buffer|.\[[ArrayBufferMaxByteLength]] to |maxsize|.
     1. Set |buffer|.\[[ArrayBufferDetachKey]] to "WebAssembly.Memory".
     1. Return |buffer|.
 </div>
@@ -695,8 +703,8 @@ which can be simultaneously referenced by multiple {{Instance}} objects. Each
 
 <div algorithm>
     The <dfn constructor for="Memory">Memory(|descriptor|)</dfn> constructor, when invoked, performs the following steps:
-    1. Let |initial| be |descriptor|["initial"].
-    1. If |descriptor|["maximum"] [=map/exists=], let |maximum| be |descriptor|["maximum"]; otherwise, let |maximum| be empty.
+    1. Let |initial| be [=?=] [$Get$](|descriptor|, "initial").
+    1. If [=?=] [$Get$](|descriptor|, "maximum") is not undefined, let |maximum| be « [=?=] [$Get$](|descriptor|, "maximum") »; otherwise, let |maximum| be an empty List.
     1. Let |memtype| be the [=memory type=] { <b>[=limits|min=]</b> |initial|, <b>[=limits|max=]</b> |maximum| }.
     1. If |memtype| is not [=valid memtype|valid=], throw a {{RangeError}} exception.
     1. Let |store| be the [=surrounding agent=]'s [=associated store=].
@@ -853,21 +861,21 @@ Each {{Table}} object has a \[[Table]] internal slot, which is a [=table address
 
 <div algorithm>
     The <dfn constructor for="Table">Table(|descriptor|, |value|)</dfn> constructor, when invoked, performs the following steps:
-    1. Let |elementType| be [=ToValueType=](|descriptor|["element"]).
+    1. Let |elementType| be [=?=] [=ToValueType=]([=?=] [$Get$](|descriptor|, "element")).
     1. If |elementType| is not a [=reftype=],
         1. [=Throw=] a {{TypeError}} exception.
-    1. Let |initial| be |descriptor|["initial"].
-    1. If |descriptor|["maximum"] [=map/exists=], let |maximum| be |descriptor|["maximum"]; otherwise, let |maximum| be empty.
+    1. Let |initial| be [=?=] [$Get$](|descriptor|, "initial").
+    1. If [=?=] [$Get$](|descriptor|, "maximum") is not undefined, let |maximum| be « [=?=] [$Get$](|descriptor|, "maximum") »; otherwise, let |maximum| be an empty List.
     1. Let |type| be the [=table type=] { <b>[=limits|min=]</b> |initial|, <b>[=limits|max=]</b> |maximum| } |elementType|.
     1. If |type| is not [=valid tabletype|valid=], throw a {{RangeError}} exception.
 
-        Note: Because tables may have up to 2<sup>32</sup> - 1 elements in validation, the checks for [=limits|min=] ≤ 2<sup>32</sup> - 1 and [=limits|max=] ≤ 2<sup>32</sup> - 1 in [=valid limits|limits validation=] cannot fail.
+    Note: Because tables may have up to 2<sup>32</sup> - 1 elements in validation, the checks for [=limits|min=] ≤ 2<sup>32</sup> - 1 and [=limits|max=] ≤ 2<sup>32</sup> - 1 in [=valid limits|limits validation=] cannot fail.
     1. If |value| is missing,
         1. Let |ref| be [=DefaultValue=](|elementType|).
     1. Otherwise,
         1. Let |ref| be [=?=] [=ToWebAssemblyValue=](|value|, |elementType|).
     1. Let |store| be the [=surrounding agent=]'s [=associated store=].
-    1. Let (|store|, |tableaddr|) be [=table_alloc=](|store|, |type|, |ref|). If allocation fails, throw a {{RangeError}} exception.
+    1. Let (|store|, |tableaddr|) be [=table_alloc=](|store|, |type|, |ref|).
     1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
     1. [=initialize a table object|Initialize=] **this** from |tableaddr|.
 </div>
@@ -885,7 +893,7 @@ Each {{Table}} object has a \[[Table]] internal slot, which is a [=table address
     1. Let |result| be [=table_grow=](|store|, |tableaddr|, |delta|, |ref|).
     1. If |result| is [=error=], throw a {{RangeError}} exception.
 
-        Note: The above exception can happen due to either insufficient memory or an invalid size parameter.
+    Note: The above exception can happen due to either insufficient memory or an invalid size parameter.
 
     1. Set the [=surrounding agent=]'s [=associated store=] to |result|.
     1. Return |initialSize|.
@@ -963,6 +971,7 @@ which can be simultaneously referenced by multiple {{Instance}} objects. Each
     1. Assert: |map|[|globaladdr|] doesn't [=map/exist=].
     1. Set |global|.\[[Global]] to |globaladdr|.
     1. [=map/Set=] |map|[|globaladdr|] to |global|.
+    1. Return ~unused~.
 </div>
 
 <div algorithm>
@@ -977,13 +986,15 @@ which can be simultaneously referenced by multiple {{Instance}} objects. Each
 
 <div algorithm>
     The algorithm <dfn>ToValueType</dfn>(|s|) performs the following steps:
-    1. If |s| equals "i32", return [=i32=].
-    1. If |s| equals "i64", return [=i64=].
-    1. If |s| equals "f32", return [=f32=].
-    1. If |s| equals "f64", return [=f64=].
-    1. If |s| equals "v128", return [=v128=].
-    1. If |s| equals "anyfunc", return [=funcref=].
-    1. If |s| equals "externref", return [=externref=].
+    1. Let |str| be [=?=] [$ToString$](|s|).
+    1. If |str| equals "i32", return [=i32=].
+    1. If |str| equals "i64", return [=i64=].
+    1. If |str| equals "f32", return [=f32=].
+    1. If |str| equals "f64", return [=f64=].
+    1. If |str| equals "v128", return [=v128=].
+    1. If |str| equals "anyfunc", return [=funcref=].
+    1. If |str| equals "externref", return [=externref=].
+    1. Throw a {{TypeError}} exception.
     1. Assert: This step is not reached.
 </div>
 
@@ -994,25 +1005,26 @@ which can be simultaneously referenced by multiple {{Instance}} objects. Each
     1. If |valuetype| equals [=f32=], return [=f32.const=] 0.
     1. If |valuetype| equals [=f64=], return [=f64.const=] 0.
     1. If |valuetype| equals [=funcref=], return [=ref.null=] [=funcref=].
-    1. If |valuetype| equals [=externref=], return [=ToWebAssemblyValue=](undefined, |valuetype|).
+    1. If |valuetype| equals [=externref=], return [=?=] [=ToWebAssemblyValue=](undefined, |valuetype|).
     1. Assert: This step is not reached.
 </div>
 
 <div algorithm>
     The <dfn constructor for="Global">Global(|descriptor|, |v|)</dfn> constructor, when invoked, performs the following steps:
-    1. Let |mutable| be |descriptor|["mutable"].
-    1. Let |valuetype| be [=ToValueType=](|descriptor|["value"]).
+    1. Let |mutable| be [=?=] [$Get$](|descriptor|, "mutable").
+    1. Let |valuetype| be [=?=] [=ToValueType=]([=?=] [$Get$](|descriptor|, "value")).
     1. If |valuetype| is [=v128=],
         1. Throw a {{TypeError}} exception.
     1. If |v| is missing,
         1. Let |value| be [=DefaultValue=](|valuetype|).
     1. Otherwise,
-        1. Let |value| be [=ToWebAssemblyValue=](|v|, |valuetype|).
+        1. Let |value| be [=?=] [=ToWebAssemblyValue=](|v|, |valuetype|).
     1. If |mutable| is true, let |globaltype| be [=var=] |valuetype|; otherwise, let |globaltype| be [=const=] |valuetype|.
     1. Let |store| be the current agent's [=associated store=].
-    1. Let (|store|, |globaladdr|) be [=global_alloc=](|store|, |globaltype|, |value|). <!-- TODO(littledan): Report allocation failure https://github.com/WebAssembly/spec/issues/584 -->
+    1. Let (|store|, |globaladdr|) be [=global_alloc=](|store|, |globaltype|, |value|).
     1. Set the current agent's [=associated store=] to |store|.
     1. [=initialize a global object|Initialize=] **this** from |globaladdr|.
+    1. Return ~unused~.
 </div>
 
 <div algorithm>
@@ -1035,7 +1047,7 @@ which can be simultaneously referenced by multiple {{Instance}} objects. Each
     1. Let |mut| |valuetype| be [=global_type=](|store|, |globaladdr|).
     1. If |valuetype| is [=v128=], throw a {{TypeError}}.
     1. If |mut| is [=const=], throw a {{TypeError}}.
-    1. Let |value| be [=ToWebAssemblyValue=](**the given value**, |valuetype|).
+    1. Let |value| be [=?=] [=ToWebAssemblyValue=](**the given value**, |valuetype|).
     1. Let |store| be [=global_write=](|store|, |globaladdr|, |value|).
     1. If |store| is [=error=], throw a {{RangeError}} exception.
     1. Set the current agent's [=associated store=] to |store|.
@@ -1100,7 +1112,7 @@ This slot holds a [=function address=] relative to the [=surrounding agent=]'s [
     1. [=list/iterate|For each=] |t| of |parameters|,
         1. If |argValues|'s [=list/size=] &gt; |i|, let |arg| be |argValues|[|i|].
         1. Otherwise, let |arg| be undefined.
-        1. [=list/Append=] [=ToWebAssemblyValue=](|arg|, |t|) to |args|.
+        1. [=list/Append=] [=?=] [=ToWebAssemblyValue=](|arg|, |t|) to |args|.
         1. Set |i| to |i| + 1.
     1. Let (|store|, |ret|) be the result of [=func_invoke=](|store|, |funcaddr|, |args|).
     1. Set the [=surrounding agent=]'s [=associated store=] to |store|.
@@ -1138,7 +1150,7 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
         1. Let |wasmValues| be a new, empty [=list=].
         1. If |values|'s [=list/size=] is not |resultsSize|, throw a {{TypeError}} exception.
         1. For each |value| and |resultType| in |values| and |results|, paired linearly,
-            1. [=list/Append=] [=ToWebAssemblyValue=](|value|, |resultType|) to |wasmValues|.
+            1. [=list/Append=] [=?=] [=ToWebAssemblyValue=](|value|, |resultType|) to |wasmValues|.
         1. Return |wasmValues|.
 </div>
 
@@ -1167,79 +1179,79 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
 <div algorithm>
 The algorithm <dfn>ToJSValue</dfn>(|w|) coerces a [=WebAssembly value=] to a JavaScript value by performing the following steps:
 
-1. Assert: |w| is not of  the form [=v128.const=] <var ignore>v128</var>.
-1. If |w| is of the form [=i64.const=] |i64|,
-    1. Let |v| be [=signed_64=](|i64|).
-    1. Return [=ℤ=](|v| interpreted as a mathematical value).
-1. If |w| is of the form [=i32.const=] |i32|, return [=𝔽=]([=signed_32=](|i32| interpreted as a mathematical value)).
-1. If |w| is of the form [=f32.const=] |f32|,
-    1. If |f32| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
-    1. If |f32| is [=nan=], return **NaN**.
-    1. Return [=𝔽=](|f32| interpreted as a mathematical value).
-1. If |w| is of the form [=f64.const=] |f64|,
-    1. If |f64| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
-    1. If |f64| is [=nan=], return **NaN**.
-    1. Return [=𝔽=](|f64| interpreted as a mathematical value).
-1. If |w| is of the form [=ref.null=] <var ignore>t</var>, return null.
-1. If |w| is of the form [=ref.func=] |funcaddr|, return the result of creating [=a new Exported Function=] from |funcaddr|.
-1. If |w| is of the form [=ref.extern=] |externaddr|, return the result of [=retrieving an extern value=] from |externaddr|.
-
-Note: Number values which are equal to NaN may have various observable NaN payloads; see [$NumericToRawBytes$] for details.
+    1. Assert: |w| is not of  the form [=v128.const=] <var ignore>v128</var>.
+    1. If |w| is of the form [=i64.const=] |i64|,
+        1. Let |v| be [=signed_64=](|i64|).
+        1. Return [=ℤ=](|v|).
+    1. If |w| is of the form [=i32.const=] |i32|, return [=𝔽=]([=signed_32=](|i32|)).
+    1. If |w| is of the form [=f32.const=] |f32|,
+        1. If |f32| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
+        1. If |f32| is [=nan=], return **NaN**.
+        1. Return [=𝔽=](|f32|).
+    1. If |w| is of the form [=f64.const=] |f64|,
+        1. If |f64| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
+        1. If |f64| is [=nan=], return **NaN**.
+        1. Return [=𝔽=](|f64|).
+    1. If |w| is of the form [=ref.null=] <var ignore>t</var>, return null.
+    1. If |w| is of the form [=ref.func=] |funcaddr|, return the result of creating [=a new Exported Function=] from |funcaddr|.
+    1. If |w| is of the form [=ref.extern=] |externaddr|, return the result of [=retrieving an extern value=] from |externaddr|.
+
+    Note: Number values which are equal to NaN may have various observable NaN payloads; see [$NumericToRawBytes$] for details.
 </div>
 
 <div algorithm>
 
 For <dfn>retrieving an extern value</dfn> from an [=extern address=] |externaddr|, perform the following steps:
 
-1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
-1. Assert: |map|[|externaddr|] [=map/exists=].
-1. Return |map|[|externaddr|].
+    1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
+    1. Assert: |map|[|externaddr|] [=map/exists=].
+    1. Return |map|[|externaddr|].
 
 </div>
 
 <div algorithm>
 The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript value to a [=WebAssembly value=] by performing the following steps:
 
-1. Assert: |type| is not [=v128=].
-1. If |type| is [=i64=],
-    1. Let |i64| be [=?=] [$ToBigInt64$](|v|).
-    1. Return [=i64.const=] |i64|.
-1. If |type| is [=i32=],
-    1. Let |i32| be [=?=] [$ToInt32$](|v|).
-    1. Return [=i32.const=] |i32|.
-1. If |type| is [=f32=],
-    1. Let |number| be [=?=] [$ToNumber$](|v|).
-    1. If |number| is **NaN**,
-        1. Let |n| be an implementation-defined integer such that [=canon=]<sub>32</sub> ≤ |n| < 2<sup>[=signif=](32)</sup>.
-        1. Let |f32| be [=nan=](n).
-    1. Otherwise,
-        1. Let |f32| be |number| rounded to the nearest representable value using IEEE 754-2019 round to nearest, ties to even mode. [[IEEE-754]]
-    1. Return [=f32.const=] |f32|.
-1. If |type| is [=f64=],
-    1. Let |number| be [=?=] [$ToNumber$](|v|).
-    1. If |number| is **NaN**,
-        1. Let |n| be an implementation-defined integer such that [=canon=]<sub>64</sub> ≤ |n| < 2<sup>[=signif=](64)</sup>.
-        1. Let |f64| be [=nan=](n).
-    1. Otherwise,
-        1. Let |f64| be |number|.
-    1. Return [=f64.const=] |f64|.
-1. If |type| is [=funcref=],
-    1. If |v| is null,
-        1. Return [=ref.null=] [=funcref=].
-    1. If |v| is an [=Exported Function=],
-        1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
-        1. Return [=ref.func=] |funcaddr|.
-    1. Throw a {{TypeError}}.
-1. If |type| is [=externref=],
-    1. If |v| is null,
-        1. Return [=ref.null=] [=externref=].
-    1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
-    1. If a [=extern address=] |externaddr| exists such that |map|[|externaddr|] is the same as |v|,
+    1. Assert: |type| is not [=v128=].
+    1. If |type| is [=i64=],
+        1. Let |i64| be [=?=] [$ToBigInt64$](|v|).
+        1. Return [=i64.const=] |i64|.
+    1. If |type| is [=i32=],
+        1. Let |i32| be [=?=] [$ToInt32$](|v|).
+        1. Return [=i32.const=] |i32|.
+    1. If |type| is [=f32=],
+        1. Let |number| be [=?=] [$ToNumber$](|v|).
+        1. If |number| is **NaN**,
+            1. Let |n| be **2147483648**<sub>ℤ</sub>.
+            1. Let |f32| be [=nan=](|n|).
+        1. Otherwise,
+            1. Let |f32| be |number| rounded to the nearest representable value using IEEE 754-2019 round to nearest, ties to even mode. [[IEEE-754]]
+        1. Return [=f32.const=] |f32|.
+    1. If |type| is [=f64=],
+        1. Let |number| be [=?=] [$ToNumber$](|v|).
+        1. If |number| is **NaN**,
+            1. Let |n| be **9223372036854775808**<sub>ℤ</sub>.
+            1. Let |f64| be [=nan=](|n|).
+        1. Otherwise,
+            1. Let |f64| be |number|.
+        1. Return [=f64.const=] |f64|.
+    1. If |type| is [=funcref=],
+        1. If |v| is null,
+            1. Return [=ref.null=] [=funcref=].
+        1. If |v| is an [=Exported Function=],
+            1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
+            1. Return [=ref.func=] |funcaddr|.
+        1. Throw a {{TypeError}}.
+    1. If |type| is [=externref=],
+        1. If |v| is null,
+            1. Return [=ref.null=] [=externref=].
+        1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
+        1. If a [=extern address=] |externaddr| exists such that |map|[|externaddr|] is the same as |v|,
+            1. Return [=ref.extern=] |externaddr|.
+        1. Let [=extern address=] |externaddr| be the smallest address such that |map|[|externaddr|] [=map/exists=] is false.
+        1. [=map/Set=] |map|[|externaddr|] to |v|.
         1. Return [=ref.extern=] |externaddr|.
-    1. Let [=extern address=] |externaddr| be the smallest address such that |map|[|externaddr|] [=map/exists=] is false.
-    1. [=map/Set=] |map|[|externaddr|] to |v|.
-    1. Return [=ref.extern=] |externaddr|.
-1. Assert: This step is not reached.
+    1. Assert: This step is not reached.
 
 </div>
 
@@ -1250,10 +1262,10 @@ WebAssembly defines the following Error classes: <dfn exception>CompileError</df
 <div algorithm="create the WebAssembly namespace object">
 When the [=namespace object=] for the {{WebAssembly}} namespace is [=create a namespace object|created=], the following steps must be run:
 
-1.  Let |namespaceObject| be the [=namespace object=].
-1.  [=list/iterate|For each=] |error| of « "CompileError", "LinkError", "RuntimeError" »,
-    1.  Let |constructor| be a new object, implementing the [=NativeError Object Structure=], with <var ignore>NativeError</var> set to |error|.
-    1.  [=!=] [$DefineMethodProperty$](|namespaceObject|, |error|,  |constructor|, false).
+    1.  Let |namespaceObject| be the [=namespace object=].
+    1.  [=list/iterate|For each=] |error| of « "CompileError", "LinkError", "RuntimeError" »,
+        1.  Let |constructor| be a new object, implementing the [=NativeError Object Structure=], with <var ignore>NativeError</var> set to |error|.
+        1.  [=!=] [$DefineMethodProperty$](|namespaceObject|, |error|,  |constructor|, false).
 
 </div>
 
