diff --git a/document/js-api/index.bs b/document/js-api/index.bs
index c00cadee..0fb44f0e 100644
--- a/document/js-api/index.bs
+++ b/document/js-api/index.bs
@@ -311,7 +311,8 @@ Note:
 
 <div algorithm>
   To <dfn>compile a WebAssembly module</dfn> from source bytes |bytes|, perform the following steps:
-    1. Let |module| be [=module_decode=](|bytes|). If |module| is [=error=], return [=error=].
+    1. Let |module| be [=module_decode=](|bytes|).
+    1. If |module| is [=error=], return [=error=].
     1. If [=module_validate=](|module|) is [=error=], return [=error=].
     1. Return |module|.
 </div>
@@ -332,7 +333,7 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
 <div algorithm>
   To <dfn>construct a WebAssembly module object</dfn> from a module |module| and source bytes |bytes|, perform the following steps:
 
-    1. Let |moduleObject| be a new {{Module}} object.
+    1. Let |moduleObject| be a [=/new=] {{Module}} object.
     1. Set |moduleObject|.\[[Module]] to |module|.
     1. Set |moduleObject|.\[[Bytes]] to |bytes|.
     1. Return |moduleObject|.
@@ -344,12 +345,12 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
     1. Let |promise| be [=a new promise=].
     1. Run the following steps [=in parallel=]:
         1. [=compile a WebAssembly module|Compile the WebAssembly module=] |bytes| and store the result as |module|.
-        1. [=Queue a task=] to perform the following steps. If |taskSource| was provided, queue the task on that task source.
-            1. If |module| is [=error=], reject |promise| with a {{CompileError}} exception.
+        1. [=Queue a task=] to perform the following steps:
+            1. If |module| is [=error=], [=reject=] |promise| with a {{CompileError}} exception.
             1. Otherwise,
                 1. [=Construct a WebAssembly module object=] from |module| and |bytes|, and let |moduleObject| be the result.
                 1. [=Resolve=] |promise| with |moduleObject|.
-    1. Return |promise|.
+    1. Return |promise|.[[Promise]].
 </div>
 
 <div algorithm>
@@ -360,9 +361,10 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
 
 <div algorithm="read-the-imports">
   To <dfn>read the imports</dfn> from a WebAssembly module |module| from imports object |importObject|, perform the following steps:
-    1. If |module|.[=imports=] [=list/is empty|is not empty=], and |importObject| is undefined, throw a {{TypeError}} exception.
+    1. Let |tuples| be [=module_imports=](|module|).
+    1. If |tuples| [=list/is empty|is not empty=], and |importObject| is undefined, throw a {{TypeError}} exception.
     1. Let |imports| be « ».
-    1. [=list/iterate|For each=] (|moduleName|, |componentName|, |externtype|) of [=module_imports=](|module|),
+    1. [=list/iterate|For each=] (|moduleName|, |componentName|, |externtype|) of |tuples|,
         1. Let |o| be [=?=] [$Get$](|importObject|, |moduleName|).
         1. If |o| [=is not an Object=], throw a {{TypeError}} exception.
         1. Let |v| be [=?=] [$Get$](|o|, |componentName|).
@@ -403,15 +405,12 @@ A {{Module}} object represents a single WebAssembly module. Each {{Module}} obje
             1. Let |externtable| be the [=external value=] [=external value|table=] |tableaddr|.
             1. [=list/Append=] |externtable| to |imports|.
     1. Return |imports|.
-
-Note: This algorithm only verifies the right kind of JavaScript values are passed.
-The verification of WebAssembly type requirements is deferred to the
-"[=instantiate the core of a WebAssembly module=]" algorithm.
+    Note: This algorithm only verifies the right kind of JavaScript values are passed. The verification of WebAssembly type requirements is deferred to the "[=instantiate the core of a WebAssembly module=]" algorithm.
 </div>
 
 <div algorithm>
   To <dfn>create an exports object</dfn> from a WebAssembly module |module| and instance |instance|, perform the following steps:
-    1. Let |exportsObject| be [=!=] [$OrdinaryObjectCreate$](null).
+    1. Let |exportsObject| be [$OrdinaryObjectCreate$](null).
     1. [=list/iterate|For each=] (|name|, |externtype|) of [=module_exports=](|module|),
         1. Let |externval| be [=instance_export=](|instance|, |name|).
         1. Assert: |externval| is not [=error=].
@@ -439,7 +438,7 @@ The verification of WebAssembly type requirements is deferred to the
         1. Assert: |status| is true.
 
         Note: the validity and uniqueness checks performed during [=WebAssembly module validation=] ensure that each property name is valid and no properties are defined twice.
-    1. Perform [=!=] [$SetIntegrityLevel$](|exportsObject|, `"frozen"`).
+    1. Perform [=!=] [$SetIntegrityLevel$](|exportsObject|, ~frozen~).
     1. Return |exportsObject|.
 </div>
 
@@ -449,6 +448,7 @@ The verification of WebAssembly type requirements is deferred to the
     1. [=Create an exports object=] from |module| and |instance| and let |exportsObject| be the result.
     1. Set |instanceObject|.\[[Instance]] to |instance|.
     1. Set |instanceObject|.\[[Exports]] to |exportsObject|.
+    1. Return ~unused~.
 </div>
 
 <div algorithm>
@@ -469,17 +469,16 @@ The verification of WebAssembly type requirements is deferred to the
     1. Let |promise| be [=a new promise=].
     1. Let |module| be |moduleObject|.\[[Module]].
     1. [=Read the imports=] of |module| with imports |importObject|, and let |imports| be the result.
-        If this operation throws an exception, catch it, [=reject=] |promise| with the exception, and return |promise|.
+    1. If |imports| is an abrupt completion, [=reject=] |promise| with the |imports|, and return |promise|.
     1. Run the following steps [=in parallel=]:
         1. [=Queue a task=] to perform the following steps:
             Note: Implementation-specific work may be performed here.
             1.  [=Instantiate the core of a WebAssembly module=] |module| with |imports|, and let |instance| be the result.
-                If this throws an exception, catch it, [=reject=] |promise| with the exception, and terminate these substeps.
+            1.  If |instance| is an abrupt completion, [=reject=] |promise| with the exception, and terminate these substeps.
             1.  Let |instanceObject| be a [=/new=] {{Instance}}.
             1.  [=initialize an instance object|Initialize=] |instanceObject| from |module| and |instance|.
-                If this throws an exception, catch it, [=reject=] |promise| with the exception, and terminate these substeps.
             1. [=Resolve=] |promise| with |instanceObject|.
-    1. Return |promise|.
+    1. Return |promise|.[[Promise]].
 </div>
 
 <div algorithm>
@@ -488,14 +487,14 @@ The verification of WebAssembly type requirements is deferred to the
     1. Let |promise| be [=a new promise=].
     1. [=Upon fulfillment=] of |promiseOfModule| with value |module|:
         1. [=asynchronously instantiate a WebAssembly module|Instantiate the WebAssembly module=] |module| importing |importObject|, and let |innerPromise| be the result.
-        1. [=Upon fulfillment=] of |innerPromise| with value |instance|.
+        1. [=Upon fulfillment=] of |innerPromise| with value |instance|:
             1. Let |result| be the {{WebAssemblyInstantiatedSource}} value «[ "{{WebAssemblyInstantiatedSource/module}}" → |module|, "{{WebAssemblyInstantiatedSource/instance}}" → |instance| ]».
             1. [=Resolve=] |promise| with |result|.
         1. [=Upon rejection=] of |innerPromise| with reason |reason|:
             1. [=Reject=] |promise| with |reason|.
     1. [=Upon rejection=] of |promiseOfModule| with reason |reason|:
         1. [=Reject=] |promise| with |reason|.
-    1. Return |promise|.
+    1. Return |promise|.[[Promise]].
 </div>
 
 <div algorithm>
@@ -545,10 +544,10 @@ interface Module {
 
 <div algorithm>
   The <dfn>string value of the extern type</dfn> |type| is
-    * "function" if |type| is of the form [=func=] <var ignore>functype</var>
-    * "table" if |type| is of the form [=table=] <var ignore>tabletype</var>
-    * "memory" if |type| is of the form [=mem=] <var ignore>memtype</var>
-    * "global" if |type| is of the form [=global=] <var ignore>globaltype</var>
+    1. Return "function" if |type| is of the form [=func=] <var ignore>functype</var>
+    1. Return "table" if |type| is of the form [=table=] <var ignore>tabletype</var>
+    1. Return "memory" if |type| is of the form [=mem=] <var ignore>memtype</var>
+    1. Return "global" if |type| is of the form [=global=] <var ignore>globaltype</var>
 </div>
 
 <div algorithm>
@@ -594,7 +593,7 @@ interface Module {
     1. Set **this**.\[[Module]] to |module|.
     1. Set **this**.\[[Bytes]] to |stableBytes|.
 
-Note: Some implementations enforce a size limitation on |bytes|. Use of this API is discouraged, in favor of asynchronous APIs.
+    Note: Some implementations enforce a size limitation on |bytes|. Use of this API is discouraged, in favor of asynchronous APIs.
 </div>
 
 <h3 id="instances">Instances</h3>
@@ -614,7 +613,7 @@ interface Instance {
     1. [=Instantiate the core of a WebAssembly module=] |module| with |imports|, and let |instance| be the result.
     1. [=initialize an instance object|Initialize=] **this** from |module| and |instance|.
 
-Note: The use of this synchronous API is discouraged, as some implementations sometimes do long-running compilation work when instantiating.
+    Note: The use of this synchronous API is discouraged, as some implementations sometimes do long-running compilation work when instantiating.
 </div>
 
 <div algorithm>
@@ -885,7 +884,7 @@ Each {{Table}} object has a \[[Table]] internal slot, which is a [=table address
     1. Let |result| be [=table_grow=](|store|, |tableaddr|, |delta|, |ref|).
     1. If |result| is [=error=], throw a {{RangeError}} exception.
 
-        Note: The above exception can happen due to either insufficient memory or an invalid size parameter.
+    Note: The above exception can happen due to either insufficient memory or an invalid size parameter.
 
     1. Set the [=surrounding agent=]'s [=associated store=] to |result|.
     1. Return |initialSize|.
@@ -1167,79 +1166,79 @@ Note: Exported Functions do not have a \[[Construct]] method and thus it is not
 <div algorithm>
 The algorithm <dfn>ToJSValue</dfn>(|w|) coerces a [=WebAssembly value=] to a JavaScript value by performing the following steps:
 
-1. Assert: |w| is not of  the form [=v128.const=] <var ignore>v128</var>.
-1. If |w| is of the form [=i64.const=] |i64|,
-    1. Let |v| be [=signed_64=](|i64|).
-    1. Return [=ℤ=](|v| interpreted as a mathematical value).
-1. If |w| is of the form [=i32.const=] |i32|, return [=𝔽=]([=signed_32=](|i32| interpreted as a mathematical value)).
-1. If |w| is of the form [=f32.const=] |f32|,
-    1. If |f32| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
-    1. If |f32| is [=nan=], return **NaN**.
-    1. Return [=𝔽=](|f32| interpreted as a mathematical value).
-1. If |w| is of the form [=f64.const=] |f64|,
-    1. If |f64| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
-    1. If |f64| is [=nan=], return **NaN**.
-    1. Return [=𝔽=](|f64| interpreted as a mathematical value).
-1. If |w| is of the form [=ref.null=] <var ignore>t</var>, return null.
-1. If |w| is of the form [=ref.func=] |funcaddr|, return the result of creating [=a new Exported Function=] from |funcaddr|.
-1. If |w| is of the form [=ref.extern=] |externaddr|, return the result of [=retrieving an extern value=] from |externaddr|.
-
-Note: Number values which are equal to NaN may have various observable NaN payloads; see [$NumericToRawBytes$] for details.
+    1. Assert: |w| is not of  the form [=v128.const=] <var ignore>v128</var>.
+    1. If |w| is of the form [=i64.const=] |i64|,
+        1. Let |v| be [=signed_64=](|i64|).
+        1. Return [=ℤ=](|v| interpreted as a mathematical value).
+    1. If |w| is of the form [=i32.const=] |i32|, return [=𝔽=]([=signed_32=](|i32| interpreted as a mathematical value)).
+    1. If |w| is of the form [=f32.const=] |f32|,
+        1. If |f32| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
+        1. If |f32| is [=nan=], return **NaN**.
+        1. Return [=𝔽=](|f32| interpreted as a mathematical value).
+    1. If |w| is of the form [=f64.const=] |f64|,
+        1. If |f64| is [=+∞=] or [=−∞=], return **+∞**<sub>𝔽</sub> or **-∞**<sub>𝔽</sub>, respectively.
+        1. If |f64| is [=nan=], return **NaN**.
+        1. Return [=𝔽=](|f64| interpreted as a mathematical value).
+    1. If |w| is of the form [=ref.null=] <var ignore>t</var>, return null.
+    1. If |w| is of the form [=ref.func=] |funcaddr|, return the result of creating [=a new Exported Function=] from |funcaddr|.
+    1. If |w| is of the form [=ref.extern=] |externaddr|, return the result of [=retrieving an extern value=] from |externaddr|.
+
+    Note: Number values which are equal to NaN may have various observable NaN payloads; see [$NumericToRawBytes$] for details.
 </div>
 
 <div algorithm>
 
 For <dfn>retrieving an extern value</dfn> from an [=extern address=] |externaddr|, perform the following steps:
 
-1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
-1. Assert: |map|[|externaddr|] [=map/exists=].
-1. Return |map|[|externaddr|].
+    1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
+    1. Assert: |map|[|externaddr|] [=map/exists=].
+    1. Return |map|[|externaddr|].
 
 </div>
 
 <div algorithm>
 The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript value to a [=WebAssembly value=] by performing the following steps:
 
-1. Assert: |type| is not [=v128=].
-1. If |type| is [=i64=],
-    1. Let |i64| be [=?=] [$ToBigInt64$](|v|).
-    1. Return [=i64.const=] |i64|.
-1. If |type| is [=i32=],
-    1. Let |i32| be [=?=] [$ToInt32$](|v|).
-    1. Return [=i32.const=] |i32|.
-1. If |type| is [=f32=],
-    1. Let |number| be [=?=] [$ToNumber$](|v|).
-    1. If |number| is **NaN**,
-        1. Let |n| be an implementation-defined integer such that [=canon=]<sub>32</sub> ≤ |n| < 2<sup>[=signif=](32)</sup>.
-        1. Let |f32| be [=nan=](n).
-    1. Otherwise,
-        1. Let |f32| be |number| rounded to the nearest representable value using IEEE 754-2019 round to nearest, ties to even mode. [[IEEE-754]]
-    1. Return [=f32.const=] |f32|.
-1. If |type| is [=f64=],
-    1. Let |number| be [=?=] [$ToNumber$](|v|).
-    1. If |number| is **NaN**,
-        1. Let |n| be an implementation-defined integer such that [=canon=]<sub>64</sub> ≤ |n| < 2<sup>[=signif=](64)</sup>.
-        1. Let |f64| be [=nan=](n).
-    1. Otherwise,
-        1. Let |f64| be |number|.
-    1. Return [=f64.const=] |f64|.
-1. If |type| is [=funcref=],
-    1. If |v| is null,
-        1. Return [=ref.null=] [=funcref=].
-    1. If |v| is an [=Exported Function=],
-        1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
-        1. Return [=ref.func=] |funcaddr|.
-    1. Throw a {{TypeError}}.
-1. If |type| is [=externref=],
-    1. If |v| is null,
-        1. Return [=ref.null=] [=externref=].
-    1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
-    1. If a [=extern address=] |externaddr| exists such that |map|[|externaddr|] is the same as |v|,
+    1. Assert: |type| is not [=v128=].
+    1. If |type| is [=i64=],
+        1. Let |i64| be [=?=] [$ToBigInt64$](|v|).
+        1. Return [=i64.const=] |i64|.
+    1. If |type| is [=i32=],
+        1. Let |i32| be [=?=] [$ToInt32$](|v|).
+        1. Return [=i32.const=] |i32|.
+    1. If |type| is [=f32=],
+        1. Let |number| be [=?=] [$ToNumber$](|v|).
+        1. If |number| is **NaN**,
+            1. Let |n| be an implementation-defined integer such that [=canon=]<sub>32</sub> ≤ |n| < 2<sup>[=signif=](32)</sup>.
+            1. Let |f32| be [=nan=](n).
+        1. Otherwise,
+            1. Let |f32| be |number| rounded to the nearest representable value using IEEE 754-2019 round to nearest, ties to even mode. [[IEEE-754]]
+        1. Return [=f32.const=] |f32|.
+    1. If |type| is [=f64=],
+        1. Let |number| be [=?=] [$ToNumber$](|v|).
+        1. If |number| is **NaN**,
+            1. Let |n| be an implementation-defined integer such that [=canon=]<sub>64</sub> ≤ |n| < 2<sup>[=signif=](64)</sup>.
+            1. Let |f64| be [=nan=](n).
+        1. Otherwise,
+            1. Let |f64| be |number|.
+        1. Return [=f64.const=] |f64|.
+    1. If |type| is [=funcref=],
+        1. If |v| is null,
+            1. Return [=ref.null=] [=funcref=].
+        1. If |v| is an [=Exported Function=],
+            1. Let |funcaddr| be the value of |v|'s \[[FunctionAddress]] internal slot.
+            1. Return [=ref.func=] |funcaddr|.
+        1. Throw a {{TypeError}}.
+    1. If |type| is [=externref=],
+        1. If |v| is null,
+            1. Return [=ref.null=] [=externref=].
+        1. Let |map| be the [=surrounding agent=]'s associated [=extern value cache=].
+        1. If a [=extern address=] |externaddr| exists such that |map|[|externaddr|] is the same as |v|,
+            1. Return [=ref.extern=] |externaddr|.
+        1. Let [=extern address=] |externaddr| be the smallest address such that |map|[|externaddr|] [=map/exists=] is false.
+        1. [=map/Set=] |map|[|externaddr|] to |v|.
         1. Return [=ref.extern=] |externaddr|.
-    1. Let [=extern address=] |externaddr| be the smallest address such that |map|[|externaddr|] [=map/exists=] is false.
-    1. [=map/Set=] |map|[|externaddr|] to |v|.
-    1. Return [=ref.extern=] |externaddr|.
-1. Assert: This step is not reached.
+    1. Assert: This step is not reached.
 
 </div>
 
@@ -1250,10 +1249,10 @@ WebAssembly defines the following Error classes: <dfn exception>CompileError</df
 <div algorithm="create the WebAssembly namespace object">
 When the [=namespace object=] for the {{WebAssembly}} namespace is [=create a namespace object|created=], the following steps must be run:
 
-1.  Let |namespaceObject| be the [=namespace object=].
-1.  [=list/iterate|For each=] |error| of « "CompileError", "LinkError", "RuntimeError" »,
-    1.  Let |constructor| be a new object, implementing the [=NativeError Object Structure=], with <var ignore>NativeError</var> set to |error|.
-    1.  [=!=] [$DefineMethodProperty$](|namespaceObject|, |error|,  |constructor|, false).
+    1.  Let |namespaceObject| be the [=namespace object=].
+    1.  [=list/iterate|For each=] |error| of « "CompileError", "LinkError", "RuntimeError" »,
+        1.  Let |constructor| be a new object, implementing the [=NativeError Object Structure=], with <var ignore>NativeError</var> set to |error|.
+        1.  [=!=] [$DefineMethodProperty$](|namespaceObject|, |error|,  |constructor|, false).
 
 </div>
 
